js/socket.ts

```diff
@@ -0,0 +1,54 @@
+import { reload } from "./browser";
+import { getMetaContent } from "./util/metadata";
+import { Socket } from "phoenix";
+import { z } from "zod";
+
+export const initSocket = () => {
+  const clientRelease = getMetaContent("release");
+  const socket = new Socket("/socket", {
+    params: {
+      release: clientRelease,
+      token: getMetaContent("guardianToken"),
+    },
+  });
+
+  socket.onError((_error, _transport, _establishedConnections) => {
+    console.warn("socket error");
+  });
+
+  socket.onOpen(() => {
+    socket
+      .channel("metadata")
+      .join()
+      .receive("ok", (rawResponse: unknown) => {
+        const { authenticated, server_release: serverRelease } =
+          MetadataJoinResponse.parse(rawResponse);
+        if (serverRelease !== clientRelease) {
+          console.warn(
+            `Server has version ${serverRelease} but we have version ${clientRelease}. Reloading.`,
+          );
+          reload();
+        }
+        if (!authenticated) {
+          console.warn("Authentication failed. Reloading.");
+          reload();
```

Comment:
This makes me nervous but I don't have a good reason for that. we're absolutely positive that websocket connections get the same authentication treatment, _always_ on the server side?

Reply:
Yeah, this is a bit risky. During testing, where I forced auth to fail, this turned into an infinite reload loop, which made the screen flash repeatedly and obnoxiously.

If there aren't any bugs in our authentication, then this is fine, but this is a particularly obnoxious result if there are any bugs, and auth problems are very precedented.

Reply:
To clarify, is it even possible that the cookie would be valid (what do we call that, the Guardian cookie?) but the token wouldn't be? Wondering if it could be useful to only redirect if the socket _and an HTTP call_ fail?

This feels like the Glides hacks again :( Perhaps it's just unavoidable for auto refresh?

Reply:
No it's not possible for the cookie to be valid and the token not to be: if you refresh, you'll get a new token based on the cookie. Well, not possible unless there are bugs, but auth is complex and bugs are possible.

You mean that if the token/socket auth failed, instead of reloading immediately, we fire off a HTTP call (which we expect to fail), and refresh. But if that request somehow succeeds, we direct to the login page? This is getting increasingly complex...

Another approach could be to store in localstorage that you just refreshed so that we can avoid refreshing more than once in a row? But again, complexity...

Reply:
I think on the whole, this approach is less hacky than glides, because we get told by the server that we're not authenticated, instead of inferring it from which requests fail.

Reply:
Gotcha, yeah, I admit the bandaids for problems we haven't actually seen are not helpful. This should be fine

Reply:
Oh, also, this specific bug we're afraid of would only be to a problem with Guardian authentication, which has been stable so far. Most auth problems come from interacting with the third party systems (keycloak+EntraID), which would cause auth to completely fail, not partially fail.